### PR TITLE
Internal load balancer

### PR DESCRIPTION
**Pull request by areed**

### CODE REVIEW

File: addons/ekco/0.11.0/README.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+   Only applicable if node purging is enabled.
+- `ekco-min-ready-worker-node-count`: Don't purge the node if it will result in less than this many ready workers
+   Only applicable if node purging is enabled.
+- `ekco-should-disable-clear-nodes`: Don't force delete pods on unreachable stuck in the terminating state
+- `ekco-disable-should-maintain-rook-storage-nodes`: Whether to maintain the list of nodes to use in the CephCluster config.
+   This also enables control of the replication factor of ceph pools, scaling up and down with the number of nodes in the cluster.
+- `ekco-disable-should-install-reboot-service`: Whether to disable the reboot service, responsible for graceful node termination.
+- `rook-version`: Version of Rook to manage
+- `rook-priority-class`: Name of priority class to add to Rook 1.0 Deployments and DaemonSets. Will be created if not found.
+- `enable-internal-load-balancer`: Run haproxy on all nodes forwarding to all API server pods
```
Comment (areed) [inline addons/ekco/0.11.0/README.md:20]: These last 4 are new

File: addons/ekco/0.11.0/Manifest [hunk 2]
```diff
@@ -0,0 +1,2 @@
+image ekco docker.io/replicated/ekco:v0.11.0
+image haproxy haproxy:2.4.2
```
Comment (areed) [inline addons/ekco/0.11.0/Manifest:2]: This image is new

File: addons/ekco/0.11.0/install.sh [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+    fi
+    EKCO_CLEAR_DEAD_NODES=true
+    if [ "$EKCO_SHOULD_DISABLE_CLEAR_NODES" = "1" ]; then
+        EKCO_CLEAR_DEAD_NODES=false
+    fi
+    if [ "$ROOK_VERSION" = "1.0.4" ]; then
+        EKCO_ROOK_PRIORITY_CLASS="node-critical"
+    fi
+
+    EKCO_ENABLE_INTERNAL_LOAD_BALANCER_BOOL=false
```
Comment (areed) [inline addons/ekco/0.11.0/install.sh:32]: This block is new

File: addons/ekco/0.11.0/install.sh [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+
+    if [ "$EKCO_SHOULD_INSTALL_REBOOT_SERVICE" = "1" ]; then
+        ekco_install_reboot_service "$src"
+    fi
+
+    if kubernetes_is_master; then
+        ekco_install_purge_node_command "$src"
+    fi
+
+    if [ "$EKCO_ENABLE_INTERNAL_LOAD_BALANCER" = "1" ]; then
```
Comment (areed) [inline addons/ekco/0.11.0/install.sh:103]: This block is new

File: addons/ekco/0.11.0/install.sh [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+        spinner_until 10 kubernetes_pod_completed $podName default
+        local phase=$(kubectl get pod $podName -ojsonpath='{ .status.phase }')
+        if [ "$phase" != "Succeeded" ]; then
+            bail "Pod $podName phase: $phase"
+        fi
+        kubectl delete pod $podName --force --grace-period=0
+    done
+}
+
+# The internal load balancer is a static pod running haproxy on every node. When joining, kubeadm
```
Comment (areed) [inline addons/ekco/0.11.0/install.sh:246]: The rest of this file is new

File: addons/ekco/0.11.0/tmpl-configmap.yaml [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    rotate_certs_check_interval: 24h
+    rotate_certs: true
+    registry_cert_namespace: kurl
+    registry_cert_secret: registry-pki
+    kurl_proxy_cert_namespace: default
+    kurl_proxy_cert_secret: kotsadm-tls
+    kotsadm_kubelet_cert_namespace: default
+    kotsadm_kubelet_cert_secret: kubelet-client-cert
+    rook_priority_class: \"$EKCO_ROOK_PRIORITY_CLASS\"
+    host_task_image: replicated/ekco:v$EKCO_VERSION
```
Comment (areed) [inline addons/ekco/0.11.0/tmpl-configmap.yaml:33]: These last 4 are new

File: addons/ekco/0.11.0/README.md [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+   Only applicable if node purging is enabled.
+- `ekco-min-ready-worker-node-count`: Don't purge the node if it will result in less than this many ready workers
+   Only applicable if node purging is enabled.
+- `ekco-should-disable-clear-nodes`: Don't force delete pods on unreachable stuck in the terminating state
+- `ekco-disable-should-maintain-rook-storage-nodes`: Whether to maintain the list of nodes to use in the CephCluster config.
+   This also enables control of the replication factor of ceph pools, scaling up and down with the number of nodes in the cluster.
+- `ekco-disable-should-install-reboot-service`: Whether to disable the reboot service, responsible for graceful node termination.
+- `rook-version`: Version of Rook to manage
+- `rook-priority-class`: Name of priority class to add to Rook 1.0 Deployments and DaemonSets. Will be created if not found.
+- `enable-internal-load-balancer`: Run haproxy on all nodes and forward to all Kubernetes API server pods
```
Comment (areed) [inline addons/ekco/0.11.0/README.md:20]: These last 3 are new

**Comment by emosbaugh** [comment]
```
Error response from daemon: manifest for replicated/ekco:v0.11.0 not found: manifest unknown: manifest unknown
```

File: addons/ekco/0.11.0/install.sh [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+    fi
+
+    # Check if load balancer is already bootstrapped
+    if curl -skf "https://localhost:6444/healthz"; then
+        return 0
+    fi
+
+    if commandExists docker; then
+        mkdir -p /etc/haproxy
+        docker run --rm -i \
```
Comment (emosbaugh) [inline addons/ekco/0.11.0/install.sh:262]: why `-i`?

File: addons/ekco/0.11.0/install.sh [hunk 9]
```diff
@@ -0,0 +1,339 @@
...
+            --entrypoint="/usr/bin/ekco" \
+            replicated/ekco:v$EKCO_VERSION \
+            generate-haproxy-config --primary-host=${backends} \
+            > /etc/haproxy/haproxy.cfg
+
+        mkdir -p /etc/kubernetes/manifests
+        docker run --rm -i \
```
Comment (emosbaugh) [inline addons/ekco/0.11.0/install.sh:269]: why `-i`?

**Review by emosbaugh — APPROVED** [review]

File: addons/ekco/0.11.0/install.sh (hunk 8 above)
Comment (areed) [reply addons/ekco/0.11.0/install.sh]: removed

**Review by emosbaugh — APPROVED** [review]

**Pull request merged**